### ISSUE TITLE
Use FECache in FEDataInterpolation

### ISSUE THREAD

**Issue by drwells** [IBAMR/IBAMR#482]
We seriously need to do this: fixing this should be easy and we still spend a ton of time recomputing shape functions.

**Comment by drwells** [comment]
Between this, the dof caching, and #444 we should be able to cut off another 15-20% of total run time, which would be nearly as good as the first round of serial performance patches.

**Comment by drwells** [comment]
Another function to tackle in this class is interpolateCommon. We call this so many times that it may be worthwhile to use some more expensive optimizations:
1. Use denser objects than `std::vector<std::vector<std::vector<std::vector<VectorValue>>>>`. We can probably use `boost::multi_array`.
2. Template the function based on the number of basis functions per cell so that we can unroll some loops.
3. Template the function on the number of variables so we can unroll more loops.
4. Make sure that computing gradients is optional.

**Comment by drwells** [comment]
I want to get to this soon but, given the C++14 issues, I think this should wait until after we push out another release.

The heart model spends about 17% of its time calling this function, most of which is spent recomputing shape function values. However, since the current version of this class handles multiple systems and finite elements its not trivial to put in FECache and initialize things correctly,

**Comment by drwells** [comment]
`FEDataInterpolation` might also be a source of excessive memory usage, as discussed in #684. A few notes:
1. The primary user of this class is `IBFE(Surface)Method::assembleInteriorForce()`: `IBFECentroidPostProcessor` also uses it. I don't think it is ever directly used in applications.
2. This class does two things:
   a. Save some computational effort for vector-valued FE fields (e.g., only compute derivatives of shape functions for one component of `X` and copy them to the others).
   b. Calculate values of the FE field at quadrature points (like [dealii::FEValues::get_function_gradients()](https://dealii.org/current/doxygen/deal.II/classFEValuesBase.html#ad1f4e0deb5d982e8172d82141c634a67)).

It can also compute values on faces. I suspect it would be clearer to split face functionality into a separate class (i.e., like [dealii::FEValues](https://dealii.org/current/doxygen/deal.II/classFEValues.html) and [dealii::FEFaceValues](https://dealii.org/current/doxygen/deal.II/classFEFaceValues.html), which have somewhat different reinitialization logic)

We can definitely improve performance by using our own cache objects inside this class (and, once #444 is in, that too).

**Comment by drwells** [comment]
Now that most of the pieces have been written (e.g., the JxW stuff) replacing this class is going to be a project for January.

**Comment by drwells** [comment]
The recent discussion about FESurfaceDistanceEvaluation is also related to this problem: we really need a class that can compute the contravariant efficiently for both the reference and deformed configuration.

**Comment by drwells** [comment]
@marshallrdavey This is the issue we discussed the other day. At this point we have written enough code that deals with changing the quadrature rule from element to element that we should be able to combine the pieces into a replacement class.

The right way to get started is to survey where we use `FECache`, `getQuadratureKey`, etc. - from there we can figure out what we should implement that would be useful in practice. For example - we frequently want to compute the positions of the quadrature points in the displaced configuration. We could combine
```cpp
Elem* const elem = patch_elems[e_idx];
const auto& X_dof_indices = X_dof_map_cache.dof_indices(elem);
get_values_for_interpolation(X_node, *X_petsc_vec, X_local_soln, X_dof_indices);

const quad_key_type key = getQuadratureKey(d_default_interp_spec.quad_type,
                                           d_default_interp_spec.quad_order,
                                           d_default_interp_spec.use_adaptive_quadrature,
                                           d_default_interp_spec.point_density,
                                           elem,
                                           X_node,
                                           patch_dx_min);
const FEBase& X_fe = X_fe_cache(key, elem);
const QBase& qrule = d_fe_data->d_quadrature_cache[key];
const std::vector<std::vector<double> >& X_phi = X_fe.get_phi();
TBOX_ASSERT(qrule.n_points() == X_phi[0].size());

for (unsigned int qp = 0; qp < qrule.n_points(); ++qp)
{
    interpolate(&X_qp[0], qp, X_node, X_phi);
    const hier::Index<NDIM> i = IndexUtilities::getCellIndex(X_qp, grid_geom, ratio);
    if (patch_box.contains(i))
    {
        (*qp_count_data)(i) += 1.0;
        ++n_local_q_points;
    }
}
```
into
```cpp
FEValues fe_values(X_system, X_petsc_vec, d_default_interp_spec, patch_dx_min);
[...]
fe_values.reinit(patch_elems[e_idx]);
const auto& X_dof_indices = X_dof_map_cache.dof_indices(elem);
const std::vector<VectorValue>& values = fe_values.get_interpolated_values(X_dof_indices);
for (unsigned int qp = 0; qp < qrule.n_points(); ++qp)
{
    const hier::Index<NDIM> i = IndexUtilities::getCellIndex(values[qp]p, grid_geom, ratio);
    if (patch_box.contains(i))
    {
        (*qp_count_data)(i) += 1.0;
        ++n_local_q_points;
    }
}
```
by putting all of the 'figure out the correct quadrature rule logic' in FEValues instead of doing it manually each time.

**Comment by drwells** [comment]
That's just one example - we should list all the things we could get this class to do before we start writing it.

**Comment by boyceg** [comment]
Could the `FEValues` class also own the DOF indexing data structures and provide some convenience function(s) to assemble into global data structures?

It seems like this is starting to become the long awaited replacement for `FEDataInterpolation`?

**Comment by drwells** [comment]
Replacing `FEDataInterpolation` is the plan!

I'm not sure if it needs to store the dof cache - it might be better to store a pointer or to rely on the user to get the dofs since its ultimately a separate job (indeed, in deal.II the `DoFCellAccessor` class extracts degrees of freedom). Putting some global update functions in it is a good idea (the more repeated code that we can move into this class the better).

**Comment by drwells** [comment]
Update: since performance elsewhere has improved a lot this class now takes up (including things it calls) 23% of the total runtime of the heart model according to valgrind.

**Comment by drwells** [comment]
I reviewed all of our FE code when I was on the plane back from Seattle the other week - I read through every use of FEDataInterpolation and also our usage of plain libMesh calculation routines. This isn't a plan of action (yet) but it is a listing of stuff that we should ultimately implement efficiently in a single place and use throughout the library.

## what FE operations do we currently do?
### libmesh_utilities.h: get_values_for_interpolation
   extract values into a `boost::multi_array`.
### libmesh_utilities.h: jacobian
   compute the Jacobian in the deformed configuration. Pseudocode:
```
fe_values.reinit(elem, X_node);
// jacobian depends on X_node
const libMesh::TypeTensor<double> &fe_values.jacobian(q_p_n) const
```
   but this function will be smart enough to compute all of these at reinit time
   and smart enough to only compute it once for triangles and tetrahedra.

   We only ever use the determinant, so
```
double fe_values.jacobian(q_p_n) const
```
   makes more sense. We might want to add something that computes the
   contravariant too but its not really important.

   It would be good to reuse code between our `JxW` implementation and this -
   they do the same thing!
### FEDataManager.cpp: sum_weighted_elem_solution
   its a nice convenience function that is also a bottleneck - we use it in two
   ways:
   1. computing force values at quadrature points (uses JxW)
   2. computing mapped quadrature points

   this is a matrix-vector multiply - perhaps it would be more
   efficient to somehow use Eigen?

   Like elsewhere, we really need to dump =std::vector<std::vector<double>>= in
   favor of something dense, possibly with compile-time determined sizes.
### FEDataManager::spread
#### In the future: deformed element bounding box
- For first order elements: just bound nodes (its good enough)
- For higher order elements: actually bound the quadrature points
#### FECache
need values of shape functions at quadrature points on reference element
- done with =sum_weighted_elem_solution=
#### JacobianCalculatorCache
need JxW values
#### QuadratureCache
need deformed quadrature points
### FEDataManager::prolongData
- deformed element bounding box
- other stuff I don't understand
### FEDataManager::interpWeighted
#### FECache
need values of shape functions at quadrature points on reference element
- done with `sum_weighted_elem_solution`
#### JacobianCalculatorCache
need JxW values in the reference configuration
#### something that efficiently adds values into ghosted vectors
its not clear to me why we evade libMesh's code here (circa line 1804 in FEDataManager.cpp)
#### RHS assembly
done with `integrate_elem_rhs`
### FEDataManager::applyGradientDetector
#### interpolate
should be replaced by `sum_weighted_elem_solution` everywhere with unit weights.
### IBFEMethod::assembleInteriorForceDensityRHS
this uses FEDataInterpolation so its harder to understand than plain libMesh data structures
- face term: normal vectors
- face term: JxW
- face term: quadrature points
- face term: shape values
- face term: solution values
- face term: solution gradients
- face term: PK1 stress tensor values (provided by a user routine)
## some ideas
### permit reinitialization with an Elem* and a quadrature key
in addition to just an Elem*. Sometimes we calculate the quadrature key, go
 do stuff, and then loop over the elements again. We can save some work by
 allowing users to provide a quadrature key that skips element shape
 calculations.
### teach FEValues how to calculate its own quadrature keys
 for convenience. Also for the case (in =IBFEMethod=) where we do not want to
 use adaptive quadrature - all quadrature keys will be the same.
### It might be nice to have our own table class
there are places where we know the size of the table (it fits a template
argument) and it would be more efficient to just read values off from the
underlying buffer instead of doing a bunch of index calculations.

Can we get access to the raw buffer stored by `boost::multi_array`? I guess
we can by just taking the address of the first entry. This might actually
work well:
```
const boost::mult_array<double, 2> &shape_values = get_values();
// n_basis_functions and n_quad_points can be template parameters
TBOX_ASSERT(shape_values.shape()[0] == n_basis_functions);
TBOX_ASSERT(shape_values.shape()[0] == n_quad_points);
const double * const _libmesh_restrict values = &shape_values[0][0];
// now access values with our own indexing logic
```
even if we don't abuse `boost::multi_array` with `restrict` we should get rid
of `std::vector<std::vector<double>>`.

**Comment by drwells** [comment]
There is another class of things that we compute inside PK1 stress functions that could be further optimized in ways that are not obvious - I just reviewed some of @jabrown893's code (the TAVR model) with valgrind and, surprisingly, we can speed up this code by 25% by getting rid of calls to `pow`. Put another way: having a class that also efficiently calculates things like `I1_bar` and `f_bar` in our solid mechanics code would be a nice quality of life improvement and a great performance boost.

**Comment by drwells** [comment]
Interesting reading: https://entropymine.com/imageworsener/slowpow/

**Comment by drwells** [comment]
The TAVR code spends about 40% of its time computing stress functions, including 10% just recomputing shape values. We can get that down significantly by making the changes listed here.

**Comment by drwells** [comment]
This project is going to take more time than I have to do it in the next two months - I think @marshallrdavey and I should tackle this over the summer. Hence I am going to remove it from the release issue list.
